### DISCUSSION TITLE
How to have api handler functions implemented on a struct to keep state?

### DISCUSSION THREAD

**Discussion by halvors** [trillium-rs/trillium#679] [Q&A]
Is it possible to pass a function implemented on a struct (self) to router? I would imagine this would work something like the code below, but i cannot get it to compile. Does anyone have any ideas on how to do this?

To clearify, i need to pass either self or Arc<RwLock<ConnectionManager>> into the handler function

```rust
#[derive(Serialize)]
struct Post {
    user_id: String,
    id: Option<usize>,
    title: String,
    body: String,
}

pub struct ManagementServer {
    connection_manager: Arc<RwLock<ConnectionManager>>,
}

impl ManagementServer {
    pub fn new(connection_manager: Arc<RwLock<ConnectionManager>>) -> Self {
        Self { connection_manager }
    }

    pub async fn listen(&self) {
        info!("Started HTTP server");

        trillium_smol::config()
            .with_host("0.0.0.0")
            .with_port(3074)
            .run_async((logger(), router().get("/", api(self.get_post))))
            .await
    }

    pub async fn get_post(&self, _conn: &mut Conn, _: ()) -> Body<Post> {
        Body(Post {
            user_id: "trillium@trillium.rs".into(),
            id: Some(10),
            title: "post title".into(),
            body: "body".into(),
        })
    }
}
```

**Comment by jbr** [comment]
Apologies for the delay; I have a broken finger so typing is harder than usual. I don't think there's a strong advantage of using trillium-api when implementing a handler on a struct, versus implementing trillium::Handler. In theory if [MutBorrowConn](https://docs.trillium.rs/src/trillium_api/api_handler.rs.html#6-11) were public, it would be possible to implement that, but I'd want to understand the use case in more detail before making that change. What do you gain by using trillium-api over implementing Handler?

Are you sure you need self in the handler? The connection manager in your example looks like conn [state](https://docs.trillium.rs/trillium/struct.state) to me.

**Comment by halvors** [comment]
@jbr Thanks for your answer, and hope your finger is healing quickly!

I think you are right in that i don't really need the api handler here, my use case here is to have a REST api where requests reads state of "ConnectionManager", so probably `https://docs.trillium.rs/trillium/struct.state` is what i was looking for.

But maybe it would be beneficial to have all the features of API handler even for this? I mean the same would be true for any other requests, i can ofc implement the same features as the API handler does myself :-)

Also do i understand it correctly that it is impossible as is to have a connection state as you mention above if using an API handler?
